More docs improvements

PeridexisErrant opened the pull request:
Now ready to merge!  Closes #670.  Closes #703.  Includes docs-in-scripts, linting, improved organisation (including automatic sorting of entries), and much more.

You can see the [docs here](http://peridexiserrant.neocities.org), and I've got  https://dfhack.rtfd.org ready to start serving the same as soon as this pull is merged - with per-branch, per-tag docs.  If someone [enables the webhook](https://read-the-docs.readthedocs.org/en/latest/webhooks.html) (takes one minute) it'll automatically update when commits are merged.

Also after this pull is merged, I'll make pulls against the 3rdparty scripts repos to put docs in those scripts.  The globbing and inclusion of their README files is a perfectly good solution for the time being though.

lethosor commented:
What's the issue with adding custom styles? I'd prefer to keep a single build system rather than have two additional build systems for different platforms that need to be maintained.

Also, I think you need to update CMakeLists.txt to point to NEWS.rst instead of NEWS. Thanks for formatting that, by the way!

PeridexisErrant commented:
Custom styles... just don't get copied across correctly, for some reason.  I'm not sure why, given that I've got it working for the DF-Walkthrough project.  It may be related to the *'dfhack-icon.ico' does not exist* errors I always get, despite it existing and being used correctly.

For build systems, all I'm proposing is to rename `conf.py.in` back to `conf.py` (so Sphinx can use it directly), and add one-line sh and bat scripts to call the docs builder.

Before merging I definitely want to get the custom styles working, and fix any cmake issues (NEWS -> NEWS.rst).

Depending on feedback I could also allow independent building of docs and/or format the `Future` section of NEWS.rst (which could risk merge issues).

Further upgrading the scripts autodoc function and moving docs into scripts would be cool, but it's also way beyond scope for this time.  Another pull, when I'm done with my current academic deadlines...

PeridexisErrant commented:
This is now ready to merge.  The only upstream change required is that a link in maxthyme README should be changed from `plugins/strangemood` to `strangemood`.

* Fixed CMake issues - or at least I can't find any remaining problems
* Configured so Sphinx can also run independently of CMake with identical results
* Added custom keybinding and in-game text styles (and used them in Plugins.rst)

It will also now work correctly on readthedocs.org - if someone [adds the webhook](https://read-the-docs.readthedocs.org/en/latest/webhooks.html), I'll set up everything else and add the DFHack team as co-owners.  This means that we can have live, auto-built documentation for every release and it natively supports [multiple versions](https://read-the-docs.readthedocs.org/en/latest/versions.html) of docs too.  Easiest online docs ever!

dscorbett commented on docs/styles/dfhack.css:
```diff
@@ -0,0 +1,40 @@
+ * Use this for any text from an in-game announcement or menu.
+ */
+    color: #CBC7C0;
+    font-family: cp437;, &#39;fixedsys&#39;, monospace;
```
This line should be:
```css
    font-family: cp437, 'fixedsys', monospace;
```

PeridexisErrant commented:
Huh, turns out the autodoc improvements worked almost the first time - [here's the branch](https://github.com/PeridexisErrant/dfhack/compare/NEWS-cleanup...PeridexisErrant:autodoc-upgrade?expand=1).  You can see the preliminary scripts documentation [here](http://peridexiserrant.neocities.org/html/docs/Scripts.html), (autogen [here]()) though obviously there's a fair bit to move over to the script files still.

This pull is still ready to merge, but if I finish before anyone the pull gets merged I'll merge the branches and include all those changes here too.

PeridexisErrant commented:
Closes #670.  Closes #703.  Includes docs-in-scripts, linting, improved organisation (including automatic sorting of entries), and much more.

You can see the [docs here](http://peridexiserrant.neocities.org), and I've got  https://dfhack.rtfd.org ready to start serving the same as soon as this pull is merged - with per-branch, per-tag docs.  If someone [enables the webhook](https://read-the-docs.readthedocs.org/en/latest/webhooks.html) (takes one minute) it'll automatically update when commits are merged.

Also after this pull is merged, I'll make pulls against the 3rdparty scripts repos to put docs in those scripts.  The globbing and inclusion of their README files is a perfectly good solution for the time being though.

PeridexisErrant commented:
After reading through all the issues one last time to make sure everything was covered, I went back and dealt with all the CMake issues I was avoiding.  Docs are now built the same way no matter what context it's in, and users can see the source if they're interested (or rebuild with updates as downstream distributors).  Relative links will always work, no only-in-one-build-system errors, and it's still correctly hooked into the CMake depends and targets.

dscorbett commented on scripts/digfort.rb:
```diff
@@ -1,4 +1,33 @@
+    d;d;d;i
+
+Available tile shapes are named after the 'dig' menu shortcuts:
+``d`` for dig, ``u`` for upstairs, ``d`` downstairs, ``i`` updown,
```
The down stair code is `j`.

PeridexisErrant replied:
Copied verbatim from the previous docs, but I'll fix that now :)

dscorbett commented on scripts/digfort.rb:
```diff
@@ -1,4 +1,33 @@
+    d;d;d;j
+
+Available tile shapes are named after the 'dig' menu shortcuts:
+``d`` for dig, ``u`` for upstairs, ``d`` downstairs, ``j`` updown,
```
The down stairs code is `j` and the up/down stairs code is `i`.

PeridexisErrant replied:
Oops.  I shouldn't edit while I'm that tired...

PeridexisErrant commented:
@lethosor @expwnent - is there anything else I need to do, or can this be merged now?

I just feel that the current docs in develop are only halfway:  they demonstrate that the system works, but it's not complete.  This pull finishes the job and the webhook makes it useful.

lethosor commented:
In #708, it was mentioned that [this section](https://github.com/PeridexisErrant/dfhack/tree/4be9e13d743f24d799582e681cc42c4f1d4f0acc#documentation) isn't complete when viewed on GitHub. Is there a way to change that?

lethosor commented on NEWS.rst:
```diff
@@ -0,0 +1,958 @@
+
+New scripts
+-----------
+- `burial`:  sets all unowned coffins to allow burial ("-pets" to allow pets too)
```
Do the single backticks serve a purpose? They seem to be ignored by Github's renderer.

PeridexisErrant commented:
Well, kinda - Github obviously doesn't use Sphinx to build docs, so much of the syntax doesn't work (eg backticks to link, or tables of contents.

The usual 'correct' solution is to have separate files for the index on sphinx and the readme on github.  This does require the docs to be built online (ie that webhook added, since I've done the rest).  I can then either add the changes to this pull, or make a new one after it's merged.  I've actually started this, so just let me know where to push it.

The readme would just be a short description of the project, links to the docs / forum thread / IRC, and maybe some of the badges for docs and testing.

lethosor commented:
My main concern is having this visible at https://github.com/DFHack/dfhack:
> DFHack documentation is generated by Sphinx. Check out the table of contents below, or the sources in the docs folder!

> User Manual:

> Other Contents:

> For Developers:

It would be nice if there's a way to only add that section with Sphinx, and maybe display something else on Github (if that's what you meant you're working on, go right ahead).

PeridexisErrant commented:
Yep, that's what it'll do - expect an additional commit later today.  Of course, some of the links won't work until it's merged!

PeridexisErrant commented:
@lethosor, to avoid multiplying merge commits I'll leave documenting the new script (#716) to you.  Just add the following in a multiline comment like other lua scripts:

    spawnunit
    =========
    ``spawnunit RACE CASTE`` creates a unit of the given race and caste at the cursor, by
    wrapping `modtools/create-unit`.  See ``spawnunit help`` for more options.

warmist commented:
could we change that a bit so it could be used the same as invoked with help?
I.e. print the same. That way we would not need to document every script twice.

PeridexisErrant commented:
Easily - if the documentation is suitable to print as the help string, just assign `helpmsg = [[=begin\n\n <msg...>`.  [view-item-info](https://github.com/PeridexisErrant/dfhack/blob/NEWS-cleanup/scripts/view-item-info.lua) does this.

This would include the `=begin` and `=end` tags though, unless they're stripped out, and rst-formatted text isn't always the best choice for CLI help.

warmist commented:
For that i suggest some simple helper function that would strip some formatting out.
The main issue here is that scripts are meant to be lightweight. If devs would need to document everything twice it would be huge pain. Even documenting it once is already annoying (as opposed to inline docs ala doxygen)

PeridexisErrant commented:
Personally I'd go all the way and have a global `help scriptname` command that understands how to parse the doc text from the file - it's pretty easy to find the string between two known tags, and process it.  If I wanted it robust I'd have to use Python, otherwise I'd include that in this pull.

Scripts *only* need to be documented in the script now though, that's the neat bit of this pull.  It's not quite as automatic as Sphinx can be for Python, but it's close.  [This text](http://peridexiserrant.neocities.org/html/_sources/docs/_auto/base.txt) is autogenerated, listing all scripts containing the correct tags and adding a link target before each.  [This page](http://peridexiserrant.neocities.org/html/docs/_auto/base.html) is the result.

Voilà, single-source docs - even if they're not totally inline.  I've also set it up to distribute the text docs, so if a user adds more (properly documented) scripts and has Python+Sphinx installed, they can update the pages from an install.  Pack authors such as myself might find that useful ;)

dscorbett commented:
The third-party script docs are broken. [dscorbett/README.html](http://peridexiserrant.neocities.org/html/scripts/3rdparty/dscorbett/README.html) shows Roses’ README and [roses/README.html](http://peridexiserrant.neocities.org/html/scripts/3rdparty/roses/README.html) shows maxthyme’s.

lethosor commented:
Does `script-in-readme.py` handle alias-only scripts, like `gui/hack-wish`? I don't see an explicit check for it, but the script only complains about `spawnunit`.

PeridexisErrant commented:
> The third-party script docs are broken. dscorbett/README.html shows Roses’ README and roses/README.html shows maxthyme’s.

Suffice to say that this was human error, compounded by free hosting and terrible cache policies.  Now fixed, I think, but either way it's not a persistent problem.

> This really should have been a separate commit

I was thinking of it as related to the readme badges, but I'll split it out.

> Py2, CMakeLists

Will do.

> Does `script-in-readme.py` handle alias-only scripts, like `gui/hack-wish`? I don't see an explicit check for it, but the script only complains about `spawnunit`.

Aliases are handled by documenting them as aliases - [here](https://github.com/PeridexisErrant/dfhack/blob/NEWS-cleanup/scripts/gui/hack-wish.lua).  It complains about spawn-unit because Travis tests the merged branches, and I haven't merged develop since `spawn-unit` was added.  I'll do that too.

PeridexisErrant commented:
I've temporarily redirected https://dfhack.readthedocs.org to build from my repo, so there's something online.  This will revert when the pull is merged and the webhook added.

5643119 moves `conf.py`, because `sphinx-build` must be run from the root dir and the host runs it in the `conf.py` directory.

dscorbett commented on docs/Plugins.rst:
```diff
@@ -698,33 +726,16 @@ dig
-~~~~~
-A permanent alias for 'digl x'.
+:digv:      Designate all of the selected vein for digging.
+:digvx:     Also cross z-levels, diggning stairs as needed.  Alias for ``digv x``.
```
“diggning” should be “digging”. Also, `diglx` should be mentioned here.

lethosor commented:
https://travis-ci.org/DFHack/dfhack/builds/88466398 is failing with `Error: Config directory doesn't contain a conf.py file.`

PeridexisErrant commented:
Ah, I'd forgotten to update CMake for the moved `conf.py`.  Done, and travis changes merged.

lethosor commented:
FWIW, Chrome is reporting certificate issues with https://dfhack.rtfd.org (http://dfhack.rtfd.org redirects to http://dfhack.readthedocs.org/en/latest/, and https://dfhack.readthedocs.org/en/latest/ works as well).

lethosor commented:
Is this ready to merge?

PeridexisErrant commented:
Yep, it's ready.  Don't forget the webhook!

That last commit is just to set up for all 3rdparty docs to also go in script source files - it's not complete yet, but working pretty well (I think it's only your scripts left).

lethosor commented:
Shouldn't the travis build be failing if my scripts aren't documented?

PeridexisErrant commented:
The linter doesn't check third party scripts yet - my plan was to get everything into shape, then make another pull to add the linter keeping it that way.

expwnent commented:
How does this compare with the [sphinx branch](https://github.com/DFHack/dfhack/tree/sphinx) I was working on a few weeks ago?

PeridexisErrant commented:
Same general concept, but taken further and with a little more polish:

- Scripts anywhere are sorted onto type-of-script based pages, regardless of where they are in the source tree ([example page](https://dfhack.readthedocs.org/en/latest/docs/_auto/modtools.html), entirely autogenerated).  Note that unlike that branch this doesn't dirty submodules, which was a pain to deal with.
- Sorting by script command and automatically added hyperlink targets
- Works at https://dfhack.readthedocs.org, with update-on-commit once the webhook is added
- Adds styles for keybindings like the wiki
- Formatted several other documents (eg NEWS, LICENSE), minor updates to others
- Adds many cross-links, subcommands, and other improvements to the docs
- Copies all the CMake docs cleanup you did and takes it a little further - build in place and copy the result, for easier debugging (because it's no longer different in each of three or four environments)
- Simple link syntax for DF bug reports (just type `:bug:'1234'` with backticks, makes a link with text "Bug 1234"), forum threads, wiki pages, etc. possible and used wherever I saw them
- Using `=begin` and `=end` for Sphinx, so Ruby scripts use native docstrings
- More natural link targets (now just the command name)
- Cleaned up Plugins.rst some more

Basically this pull includes script docs in source files, plus every little thing I could think of that I wanted the docs to do.

PeridexisErrant commented:
I've also [clarified LICENSE.rst](https://github.com/PeridexisErrant/dfhack/blob/license-clarify/LICENSE.rst), but unless there's massive support I'll leave that to a follow-up pull - too easy to look like important changes are sneaking through, even without any legal impact at all.

PeridexisErrant commented:
@expwnent @lethosor @warmist - I'm finished here, it's demonstrably working, and every travis build but this one is breaking.  Can we please merge it and move on, or is there actually something that needs to change?

Of course [adding the RTD service](https://read-the-docs.readthedocs.org/en/latest/webhooks.html) will take about 30 seconds, and enable live updates of the development version of the docs whenever someone pushes commits to `develop` - and the stable docs when commits are pushed to `master`.  I've got the backend set up already, and will just need to point it at the DFHack repo instead of mine and change the default version to `stable` (ie master) docs once they exist.

lethosor commented:
I can't do anything about the webhook since I'm not an owner, so that'll have to be done by someone else (@expwnent?).
I also just ran into these when building the docs (and cmake's configuring stage took about 2 minutes beforehand, although that could be a disk cache issue):
```
/<dfhack path>/scripts/include-all.rst:1: SEVERE: Problems with "include" directive path:
InputError: [Errno 2] No such file or directory: '../<dfhack path>/scripts/add-thought.lua'.
/<dfhack path>/scripts/devel/include-all.rst:: WARNING: document isn't included in any toctree
/<dfhack path>/scripts/fix/include-all.rst:: WARNING: document isn't included in any toctree
/<dfhack path>/scripts/gui/include-all.rst:: WARNING: document isn't included in any toctree
/<dfhack path>/scripts/include-all.rst:: WARNING: document isn't included in any toctree
/<dfhack path>/scripts/modtools/include-all.rst:: WARNING: document isn't included in any toctree
```
Any idea what the issue with the add-thought.lua is?

lethosor commented:
The webhook is enabled now.

PeridexisErrant commented:
Good to hear about the webhook :)

That error log looks like you've got all the files generated by expwnent's sphinx branch around, so git clean might be a good idea.  Or maybe a fresh clone, because `scripts/add-thought.lua` certainly exists in the pull!

lethosor commented:
Yeah, turns out `scripts/include-all.rst` had the problematic add-thought path. CMake is still taking an unusually long time for configuration, though.

lethosor commented on CMakeLists.txt:
```diff
@@ -204,43 +201,29 @@ if (BUILD_DOCS)
+    "${CMAKE_CURRENT_SOURCE_DIR}/scripts/about.txt"
+    "${CMAKE_CURRENT_SOURCE_DIR}/scripts/*/about.txt"
+  )
+  file(GLOB_RECURSE SPHINX_SCRIPT_DEPS scripts "${CMAKE_CURRENT_SOURCE_DIR}"
```
Changing this line to
```cmake
  file(GLOB_RECURSE SPHINX_SCRIPT_DEPS "${CMAKE_CURRENT_SOURCE_DIR}/scripts"
```
brought the configuration time from around 2-3 minutes to 15-20 seconds for me, so I'm pretty sure what you have here is scanning the entire DFHack source tree.

lethosor replied:
Actually, I think this would be better:
```cmake
  file(GLOB_RECURSE SPHINX_SCRIPT_DEPS
    "${CMAKE_CURRENT_SOURCE_DIR}/scripts/*.lua"
    "${CMAKE_CURRENT_SOURCE_DIR}/scripts/*.rb"
  )
```

lethosor commented:
I can merge this and make that change if you don't have anything else to add.

PeridexisErrant commented:
Yep, merge it.

Anything else can come in a later pull if I think of it.

PeridexisErrant commented:
Party time!  I've switched https://dfhack.readthedocs.org to track this repo, and with the webhook enabled any addition to `develop` will be immediately and automatically reflected in the docs.

I'll be back after the next release to make the default version "stable", but until `master` has the Sphinx documentation it just shows an error page.

If anyone with push access to DFHack has a readthedocs account, I'd be happy to add you as an admin to the docs page.
